### DISCUSSION TITLE
Have role in service user's jwt

### DISCUSSION THREAD

**Discussion by lautaroalvarez** [zitadel/zitadel#6023] [Q&A] [answered]
I have a Zitadel instance with a configured project and users.

In this project I have the "role" inside the access_tokens (checking this option)
![Captura de pantalla de 2023-06-13 08-53-15](https://github.com/zitadel/zitadel/assets/14188347/ecde04b2-5775-4b94-8c84-a5b3d6998511)

Everything is fine with user's token. I can decode it, validate with zitadel keys and obtain the roles inside the decoded token. For example:
```
{
  "iss": "https://id.myinstance.com.ar",
  "sub": "208118479318941819",
  "aud": [
    "208118419525009531@myproject",
    "208118068998570107"
  ],
  "jti": "218253358996194313",
  "exp": 1686628339,
  "iat": 1686585139,
  "nbf": 1686585139,
  "urn:zitadel:iam:org:project:roles": {
    "user": {
      "208117277516628091": "test.id.myinstance.com.ar"
    }
  }
}
``` 

I have to integrate another service that have to make request to my API. Reading the documentation I have created a service user and assign it the "external-app" role. But when I generate a jwt (using private key jwt https://zitadel.com/docs/guides/integrate/private-key-jwt) I can't get the roles inside the jwt.

For example:
```
{
  iss: 'https://id.myinstance.com.ar',
  sub: '217503117799850767',
  aud: [ 'https://id.myinstance.com.ar' ],
  jti: '218373926881854223',
  exp: 1686700203,
  iat: 1686657003,
  nbf: 1686657003
}
```

Is there any way to include roles inside my service user's jwt? I don't want to change my token validation inside API because it's really fast in this way (without making a request to zitadel instance)

**Comment by fforootd** [comment]
Hm this should work since 2.23.0 https://github.com/zitadel/zitadel/issues/5502

What version are you using? Or are you sending along the scope mentioned in the PR?

**Reply by lautaroalvarez** [reply]
Hi, I'm using image `ghcr.io/zitadel/zitadel:v2.21.1`. I will upgrade and then comment here.

**Reply by lautaroalvarez** [reply]
I have upgraded to v2.27.2 and I have the same result.

Here is my request to generate the access_token:
```
function run(privToken) {
    return axios.request({
        method: 'POST',
        url: 'https://id.myinstance.com.ar/oauth/v2/token',
        headers: {'content-type': 'application/x-www-form-urlencoded'},
        data: {
            grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
            scope: 'openid profile email urn:zitadel:iam:org:projects:roles',
            assertion: privToken
        }
    })
        .then((response) => {
            console.log(response.data);
        })
        .catch((error) => {
            console.error(error);
        });
}
```
where `privToken` is the jwt generated with the service user's private key json file.

**Reply by fforootd** [reply]
Hm that looks about right. Maybe @livio-a can spot the problem.

**Reply by hifabienne** [reply, marked as answer]
Yan you try sending the audience scope to define for which application you want to get the roles?
Replace the projectid with the id of the project you want to get the roles: urn:zitadel:iam:org:project:id:{projectid}:aud

**Reply by lautaroalvarez** [reply]
Wow! This works.

I use this function:
```
function run(privToken) {
    return axios.request({
        method: 'POST',
        url: 'https://id.myinstance.com.ar/oauth/v2/token',
        headers: {'content-type': 'application/x-www-form-urlencoded'},
        data: {
            grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
            scope: 'openid profile email urn:zitadel:iam:org:project:id:208118068998570107:aud urn:zitadel:iam:org:projects:roles',
            assertion: privToken
        }
    })
        .then((response) => {
            console.log(response.data);
        })
        .catch((error) => {
            console.error(error);
        });
}
```

And I get this jwt (decoded):
```
{
  aud: [ 'https://id.myinstance.com.ar', '208118068998570107' ],
  exp: 1686785044,
  iat: 1686741844,
  iss: 'https://id.myinstance.com.ar',
  jti: '218516266594338143',
  nbf: 1686741844,
  sub: '217503117799850767',
  'urn:zitadel:iam:org:project:208118068998570107:roles': {
    'external-app': { '208117277516628091': 'test.id.myinstance.com.ar' }
  }
}
```

Thank you both for the help!

**Reply by fforootd** [reply]
Great, let us know how your journey goes 😁
